server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java

```diff
@@ -0,0 +1,152 @@
+                submitHealthMetadata("health-metadata-update-master-election");
+            }
+            // If the node is not the elected master anymore
+            publishedAfterElection = event.localNodeMaster();
+        }
+    }
+
+    private void updateHealthMetadataIfNecessary() {
+        ClusterState clusterState = clusterService.state();
+        if (clusterState.nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_4_0)
+            && clusterState.nodes().isLocalNodeElectedMaster()) {
+            submitHealthMetadata("health-metadata-update-threshold-change");
+        }
+    }
+
+    public static List<NamedXContentRegistry.Entry> getNamedXContentParsers() {
+        return List.of(
+            new NamedXContentRegistry.Entry(Metadata.Custom.class, new ParseField(HealthMetadata.TYPE), HealthMetadata::fromXContent)
+        );
+    }
+
+    public static List<NamedWriteableRegistry.Entry> getNamedWriteables() {
+        return List.of(
+            new NamedWriteableRegistry.Entry(Metadata.Custom.class, HealthMetadata.TYPE, HealthMetadata::new),
+            new NamedWriteableRegistry.Entry(NamedDiff.class, HealthMetadata.TYPE, HealthMetadata::readDiffFrom)
+        );
+    }
+
+    private HealthMetadata createHealthMetadata() {
+        return new HealthMetadata(HealthMetadata.DiskThresholds.createDiskThresholds(diskThresholdSettings));
+    }
+
+    private void submitHealthMetadata(String source) {
+        HealthMetadata localHealthMedata = createHealthMetadata();
+        var task = new UpdateHealthMetadataTask(localHealthMedata);
+        var config = ClusterStateTaskConfig.build(Priority.NORMAL);
+        clusterService.submitStateUpdateTask(source, task, config, taskExecutor);
+    }
+
+    record UpdateHealthMetadataTask(HealthMetadata healthMetadata) implements ClusterStateTaskListener {
+
+        @Override
+        public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+            assert false : "never called";
+        }
+
+        @Override
+        public void onFailure(@Nullable Exception e) {
+            logger.error("failure during health metadata update", e);
+        }
+
+        static class Executor implements ClusterStateTaskExecutor<UpdateHealthMetadataTask> {
+
+            @Override
+            public ClusterState execute(ClusterState currentState, List<TaskContext<UpdateHealthMetadataTask>> taskContexts)
+                throws Exception {
+                final HealthMetadata initialHealthMetadata = HealthMetadata.getHealthCustomMetadata(currentState);
+                HealthMetadata currentHealthMetadata = initialHealthMetadata;
+                for (TaskContext<UpdateHealthMetadataTask> taskContext : taskContexts) {
+                    currentHealthMetadata = taskContext.getTask().healthMetadata();
+                    taskContext.success(() -> {});
+                }
+                final var finalHealthMetadata = currentHealthMetadata;
```

Comment:
is the `finalHealthMetadata` needed still?

Would checking `return currentHealthMetadata == initialHealthMetadata` below be enough?

Reply:
It's needed because of the lambda, the `currentHealthMetadata` is not final.

Reply:
Ah you're right, we can't make it effectively final due to the loop :sadpanda: 
Sorry for the noise

Reply:
Indeed, I could get the last one like we did before and make it final, but I like that this is more inline with code in other similar cases.